src/RevCPU.cc

```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
 
 // Initializes a RevThread object.
 // - Moves it to the 'Threads' map
-// - Adds it's ThreadID to the ThreadQueue to be scheduled
-void RevCPU::InitThread(std::shared_ptr<RevThread>& ThreadToInit){
-
+// - Adds it's ThreadID to the ReadyThreads to be scheduled
+void RevCPU::InitThread(std::unique_ptr<RevThread> ThreadToInit){
+  // @Lee, can this be a reference? I'm not looking to transfer ownership of the regfile pointer
+  std::unique_ptr<RevRegFile> RegState = ThreadToInit->TransferVirtRegState();
   auto gp = Loader->GetSymbolAddr("__global_pointer$");
-  ThreadToInit->GetRegFile()->SetX(RevReg::gp, gp);
-  ThreadToInit->GetRegFile()->SetX(RevReg::s0, gp);
-
-  uint32_t TID = ThreadToInit->GetThreadID();
-  // Check if this ThreadID has already been assigned... if so... something has gone horribly wrong
-  // print out all threads
-  auto it = Threads.find(TID);
-  if( it != Threads.end() && it->second->GetState() != ThreadState::START ){
-    output.fatal(CALL_INFO, 99, "Error: ThreadID %" PRIu32 " has already been assigned... this is a bug.\n", TID);
-  }
-  output.verbose(CALL_INFO, 4, 0, "Initializing Thread %" PRIu32 "\n", TID);
+  RegState->SetX(RevReg::gp, gp);
+  RegState->SetX(RevReg::s0, gp); // This can be used as the fp reg per ABI
+  output.verbose(CALL_INFO, 4, 0, "Initializing Thread %" PRIu32 "\n", ThreadToInit->GetID());
   output.verbose(CALL_INFO, 11, 0, "Thread Information: %s", ThreadToInit->to_string().c_str());
   ThreadToInit->SetState(ThreadState::READY);
-  Threads.emplace(ThreadToInit->GetThreadID(), ThreadToInit);
-  ThreadQueue.emplace_back(TID);
+  ReadyThreads.emplace_back(std::move(ThreadToInit));
 }
 
-void RevCPU::AssignThread(uint32_t ThreadID, uint32_t ProcID){
-  output.verbose(CALL_INFO, 4, 0, "Assigning Thread %" PRIu32 " to Processor %" PRIu32 "\n", ThreadID, ProcID);
-  auto Thread = Threads.at(ThreadID);
-
-  // Point the regfile of this thread's LSQ to the Proc's LSQ
-  Thread->GetRegFile()->SetLSQueue( Procs[ProcID]->GetLSQueue() );
-
-  // Point thread's regfile to this proc's MarkLoadComplete
-  Thread->GetRegFile()->SetMarkLoadComplete([proc = Procs[ProcID]](const MemReq& req){ proc->MarkLoadComplete(req); });
-
-  // Put the thread in the Proc's assigned threads list
-  AssignedThreads.at(ProcID).emplace(ThreadID, Thread);
-
-  Procs[ProcID]->AssignThread(ThreadID);
+// Assigns a RevThred to a specific Proc which then loads it into a RevHart
+// This should not be called without first checking if the Proc has an IdleHart
+void RevCPU::AssignThread(std::unique_ptr<RevThread> ThreadToAssign, unsigned ProcID){
+  output.verbose(CALL_INFO, 4, 0, "Assigning Thread %" PRIu32 " to Processor %" PRIu32 "\n", ThreadToAssign->GetID(), ProcID);
+  Procs[ProcID]->AssignThread(std::move(ThreadToAssign));
```

Comment:
In this case, you are transferring ownership of the `ThreadToAssign` to a `RevProc`. So it will not be used again until the `std::unique_ptr<RevThread>` is returned via the `ThreadsThatChangedState` handling. Does this affect your comment?

Reply:
I don't know what comment "above" I was referring to, but I think it involves the rvalue reference and move stuff, so you should change the function to:
```c++
void RevCPU::AssignThread(std::unique_ptr<RevThread>&& ThreadToAssign, unsigned ProcID){
```
The rest of the function stays the same. You are avoiding creating a temporary local `ThreadToAssign` and are using a rvalue reference to the replacement directly.